module/module_test.go

```diff
@@ -237,6 +237,23 @@ func TestModuleFunctions(t *testing.T) {
 			`error validating resource: expected "motorL" attribute for mybase "mybase2"`)
 	})
 
+	t.Run("ConvertedAttributes set correctly", func(t *testing.T) {
+		// mybase.Reconfigure attempts to access ConvertedAttributes. Ensure that
+		// Adding and Reconfiguring mybase does not result in error (both
+		// module.AddResource and module.ReconfigureResource should fill in
+		// ConvertedAttributes).
+		deps := []string{motor.Named("motor1").String(), motor.Named("motor2").String()}
+		_, err = m.AddResource(ctx, &pb.AddResourceRequest{
+			Config: myBaseConf, Dependencies: deps,
+		})
+		test.That(t, err, test.ShouldBeNil)
+
+		_, err = m.ReconfigureResource(ctx, &pb.ReconfigureResourceRequest{
```

Comment:
[q] Should this test pass in a different config (rather than the same config as AddResource had) & confirm that the well-typed value from the new config is actually set on the struct?

Reply:
While `AddResource` and `ReconfigureResource` take in the same protobuf component config, they each create different `config.Component` instances from it with different `ConvertedAttributes` fields ([here](https://github.com/viamrobotics/rdk/blob/main/module/module.go#L279) is where `AddResource` does it, e.g.). In other words, `ReconfigureResource` will not reuse the create `ConvertedAttributes` from `AddResource`.

Reply:
But shouldn't we be able to prove that from the tests alone w/o needing to look at the implementation? Also, while there is a new instance of `config.Component` being created by both `AddResource` and `ReconfigureResource` the config is semantically the same which I think is confusing b/c the main purpose (correct me if I'm wrong) of reconfigure is to reconfigure the robot on (semantic) config change. 

Proving that the attrConfig is actually changed by reconfiguration would make the desired behavior obvious by just looking at the tests.

Reply:
I agree with Nick here. This should change the config, and then actually test that the attributes are converted AND that they contain the newly reconfigured value(s).

As it is now, this actually isn't testing anything new. You could remove the changes in module.go and this would still pass, since it's never actually checking the conversion, just that the overall reconfigure didn't error (which is already true in the previous reconfigure test.)

Reply:
Hmm I'm a little lost. TL;DR I see why this test is not obvious in what it's testing, but I do think it covers the correct code paths in `module/module.go`.

> You could remove the changes in module.go and this would still pass.

Removing the changes to `module.go` and running this test fails with `'expected *mybase.MyBaseConfig but got <nil>'` (similar to the error you were originally seeing @nicksanford). I think the undesired behavior currently is that `ReconfigureResource` in the module does not set `ConvertedAttributes` with `addConvertedAttributes`. Both the [constructor function](https://github.com/viamrobotics/rdk/blob/main/examples/customresources/models/mybase/mybase.go#L46) and the [reconfiguration function](https://github.com/viamrobotics/rdk/blob/main/examples/customresources/models/mybase/mybase.go#L52) for mybase try to access `ConvertedAttributes`, so trying to `AddResource` or `ReconfigureResource` a mybase without setting converted attributes will cause the `'expected *mybase.MyBaseConfig but got <nil>'` error.

> and then actually test that the attributes are converted AND that they contain the newly reconfigured value(s)

`ConvertedAttributes` and `MyBaseConfig` are not really reachable from here, so I don't think I can assert anything about their structure. I could move this test to `manager_test` and ensure that a `mybase` can be correctly reconfigured to rely on different motors, would that be better?

Reply:
I'll pull this branch down & see if I can find an easier way to test what I'd like to test. Give me 30 ish min.

Reply:
Sorry for my poor explanation. 

My point was that the test add appears to only be testing that reconfigure returns without error. It doesn't look like its actually testing what it says (that is, that the converted attributes are "set correctly.") As there's a reconfigure test right above this already, it looked like you were simply duplicating that test. I'm honestly a little confused now how that test was passing before this fix then. Does gizmo just not have a validator?